java/samples/semantickernel-concepts/README.md

```diff
@@ -0,0 +1,436 @@
+
+```java
+public static class Time {
+
+    @DefineKernelFunction(name = "date")
+    public String date() {
+        System.out.println("date is callled");
+        Date now = new Date();
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
+        return dateFormat.format(now);
+    }
+
+    @DefineKernelFunction(name = "time")
+    public String time() {
+        System.out.println("time is called");
+        Date now = new Date();
+        SimpleDateFormat timeFormat = new SimpleDateFormat("HH:mm:ss");
+        return timeFormat.format(now);
+    }
+}
+
+KernelPlugin time = KernelPluginFactory.createFromObject(new Time(), "time");
+```
+
+## How to Define a KernelFunction?
+
+### Native function
+
+Define a local class and its function, and annotate it appropriately
+
+```java
+public class TextPlugin {
+    @DefineKernelFunction(description = "Change all string chars to uppercase.", name = "Uppercase")
+    public String uppercase(@KernelFunctionParameter(description = "Text to uppercase", name = "input") String text) {
+        return text.toUpperCase(Locale.ROOT);
+    }
+}
+```
+
+### Inline function
+
+Define a function using a Prompt.
+
+#### `KernelFunctionFromPrompt.builder().withTemplate(promptTemplate).build();`
+
+```java
+String promptTemplate = """
+    Generate a creative reason or excuse for the given event.
+    Be creative and be funny. Let your imagination run wild.
+
+    Event: I am running late.
+    Excuse: I was being held ransom by giraffe gangsters.
+
+    Event: I haven't been to the gym for a year
+    Excuse: I've been too busy training my pet dragon.
+
+    Event: {{$input}}
+""".stripIndent();
+
+var excuseFunction = KernelFunctionFromPrompt.builder()
+    .withTemplate(promptTemplate)
+    .withDefaultExecutionSettings(
+        PromptExecutionSettings.builder()
+            .withTemperature(0.4)
+            .withTopP(1)
+            .withMaxTokens(500)
+            .withUser("bx-h")
+            .build()
+    )
+    .withName("ExcuseGeneratorFunction")
+    .build();
+```
+
+#### `KernelFunction.createFromPrompt(message).build();`
+
+```java
+var message = "Translate this date " + date + " to French format";
+var fixedFunction = KernelFunction
+        .createFromPrompt(message)
+        .withDefaultExecutionSettings(
+                PromptExecutionSettings.builder()
+                        .withMaxTokens(100)
+                        .build())
+        .withTemplateFormat(PromptTemplateConfig.SEMANTIC_KERNEL_TEMPLATE_FORMAT)
+        .withName("translator")
+        .build();
+```
+
+> `KernelFunction.createFromPrompt(message)` is entirely equivalent to `KernelFunctionFromPrompt.builder().withTemplate(promptTemplate)` according to the source code.
+
+> `SEMANTIC_KERNEL_TEMPLATE_FORMAT` refers to '**semantic-kernel**', which is a rendering engine for `{{$variable}}`
+> 
+> Another rendering engine is **'handlebars'**, which is used for `{{variable}}`
```

Comment:
Can link to:

https://devblogs.microsoft.com/semantic-kernel/using-handlebars-planner-in-semantic-kernel/

and

https://learn.microsoft.com/en-us/semantic-kernel/prompts/prompt-template-syntax